setting log color from flutter log setting page

quangson91 opened the pull request:
Issue: https://github.com/flutter/flutter-intellij/issues/2396

setting log color by taking value from flutter log color setting page.

Demo
![flutter_log_color](https://user-images.githubusercontent.com/6185205/41732640-b7be77ac-75ab-11e8-8b3b-c670e697516e.gif)


/ cc @pq

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -35,8 +42,28 @@
+  }
+
+  @NotNull
+  private final EditorColorsScheme globalEditorColorsScheme = EditorColorsManager.getInstance().getGlobalScheme();
```
NIT: perhaps make this static and rename to `GLOBAL_EDITOR_COLORS_SCHEME` or maybe just inline?

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -45,20 +72,26 @@
+      }
+      catch (Exception e) {
+        // Should never go here.
+        e.printStackTrace();
```
Let's add a logger and log this instead of printing a stack trace.

Something like:

    private static final Logger LOG = Logger.getInstance(FlutterLogView.class);

and then:

    LOG.error(e);

?

quangson91 replied:
I updated.
Thanks!

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -45,20 +72,26 @@
+        // Should never go here.
+        e.printStackTrace();
+      }
+      return SimpleTextAttributes.GRAY_ATTRIBUTES;
```
Maybe default to `REGULAR_ATTRIBUTES`?

quangson91 replied:
Yeah, I updated.

quangson91 commented:
@pq I have addressed your comments.
Please take a look!.

Also, I'm working on keep user preference setting.
(Mean, keep user setting filter eg regex/match case/show timestamp/show color entry, etc.)
The code almost finished, now I'm testing.
So I will create a new PR soon.

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -352,7 +358,9 @@ String countString(int count) {
     this.app = app;
-
+    ApplicationManager.getApplication().getMessageBus().connect(this).subscribe(EditorColorsManager.TOPIC, scheme -> {
+      globalEditorColorsScheme = EditorColorsManager.getInstance().getGlobalScheme();
```
Got it.  Thanks!

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -45,20 +78,26 @@
+      try {
+        final TextAttributesKey key = LOG_LEVEL_TEXT_ATTRIBUTES_KEY_MAP.get(level);
+        final Color color = globalEditorColorsScheme.getAttributes(key).getForegroundColor();
+        return new SimpleTextAttributes(SimpleTextAttributes.STYLE_PLAIN, color);
```
Does this create a new `SimpleTextAttributes` object for each entry?  If so, would it make sense to cache them?

quangson91 replied:
Ah, you right.
We should be caching it. So we don't have re-created new `SimpleTextAttributes` value.
I implemented in this commit: https://github.com/flutter/flutter-intellij/pull/2399/commits/72ab8ff2668a61c0ff07747e2bca5dde782078ba

Please take a look!

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -45,20 +82,32 @@
+
+        final SimpleTextAttributes result = new SimpleTextAttributes(SimpleTextAttributes.STYLE_PLAIN, color);
+        textAttributesByLogLevelCached.put(level, result);
+        return result;
```
I may be misreading, but does this ever get things from the cache?

pq replied:
Also, how would a map of levels to colors get updated when `globalEditorColorsScheme` changes?

Another approach might be to keep a cache of colors to attributes...

For example:

```java
Map<Color, SimpleTextAttributes> attributeCache = new ConcurrentHashMap<>(); 

...

SimpleTextAttributes result = attributeCache.get(color); // or containsKey(...)
if (result == null) {
  result = ...
  attributeCache.put(color, result);
}
return result;
```

Then we don't have to worry about level cache consistency.

quangson91 replied:
I think we have to cache: `Map<Level, SimpleTextAttributes> textAttributesByLogLevelCache`.

Reason:
Cached values is used in method:
```java
public SimpleTextAttributes style(@Nullable FlutterLogEntry entry, int attributes) {
```

=> input
```java
final FlutterLog.Level level = FlutterLog.Level.forValue(entry.getLevel());
```
=> Ouput
```java
SimpleTextAttributes
```

=> `Map<Level, SimpleTextAttributes>` is right choise.

quangson91 replied:
When global color scheme changed -> we simply reset cache data.

https://github.com/flutter/flutter-intellij/pull/2399/commits/72ab8ff2668a61c0ff07747e2bca5dde782078ba#diff-2a35badbf1bca121cb7f36039ac67315R372

quangson91 replied:
> I may be misreading, but does this ever get things from the cache?

Yes, I tested. I'm sure it will get value from cache.

pq commented on src/io/flutter/logging/FlutterLogView.java:
```diff
@@ -35,8 +43,37 @@
+  @NotNull
+  private volatile EditorColorsScheme globalEditorColorsScheme = EditorColorsManager.getInstance().getGlobalScheme();
+  @NotNull
+  private final Map<Level, SimpleTextAttributes> textAttributesByLogLevelCached = new ConcurrentHashMap<>();
```
Maybe consider renaming to `textAttributesByLogLevelCache` ?

quangson91 replied:
Done.

quangson91 commented:
@pq, I already refactoring logic storing text-attributes.
Please see this commit: https://github.com/flutter/flutter-intellij/pull/2399/commits/91b3c7be6b49be6744ed4d23bfca38d9904d1d1d
